pyomo/core/base/units_container.py

```diff
@@ -0,0 +1,1335 @@
+    """
+    def __init__(self):
+        """Create a PyomoUnitsContainer instance. """
+        # Developers: Do not interact with this attribute directly, but instead
+        # access through the property _pint_registry since that is where the import
+        # of the 'pint' module is checked
+        self.__pint_registry = None
+
+    @property
+    def _pint_registry(self):
+        """ Return the pint.UnitsRegistry instance corresponding to this container. """
+        if pint_module is None:
+            # pint was not imported for some reason
+            raise RuntimeError("The PyomoUnitsContainer in the units_container module requires"
+                              " the package 'pint', but this package could not be imported."
+                              " Please make sure you have 'pint' installed.")
+
+        if self.__pint_registry is None:
+            self.__pint_registry = pint_module.UnitRegistry()
+
+        return self.__pint_registry
+
+    def __getattr__(self, item):
+        """
+        Here, __getattr__ is implemented to automatically create the necessary unit if
+        the attribute does not already exist.
+
+        Parameters
+        ----------
+        item : str
+            the name of the new field requested
+
+        Returns
+        -------
+        : PyomoUnit
+           returns a PyomoUnit corresponding to the requested attribute,
+           or None if it cannot be created.
+
+        """
+        # since __getattr__ was called, we must not have this field yet
+        # try to build a unit from the requested item
+        pint_registry = self._pint_registry
+        try:
+            pint_unit = getattr(pint_registry, item)
+
+            if pint_unit is not None:
+                # check if the unit is an offset unit and throw an exception if necessary
+                # ToDo: should we prevent delta versions: delta_degC and delta_degF as well?
+                pint_unit_container = pint_module.util.to_units_container(pint_unit, pint_registry)
+                for (u, e) in six.iteritems(pint_unit_container):
+                    if not pint_registry._units[u].is_multiplicative:
+                        raise UnitsError('Pyomo units system does not support the offset units "{}".'
+                                         ' Use absolute units (e.g. kelvin instead of degC) instead.'
+                                         ''.format(item))
+
+                unit = _PyomoUnit(pint_unit, pint_registry)
+                setattr(self, item, unit)
+                return unit
+        except pint_module.errors.UndefinedUnitError as exc:
+            pint_unit = None
+
+        if pint_unit is None:
+            raise AttributeError('Attribute {0} not found.'.format(str(item)))
+
+    def create_PyomoUnit(self, pint_unit):
+        return _PyomoUnit(pint_unit, self._pint_registry)
+
+    # ToDo: Add support to specify a units definition file instead of this programatic interface
```

Comment:
"TODO"  (todo's in all caps are actually picked up by the Jenkins builds and catalogued)

Reply:
I fixed this, but there appears to be some auto-switching of TODO to .. todo:: that is messing up the docstrings. Therefore, I also moved my long todo list outside the docstring.